# Key-Recovery-Attack Friendly Distinguisher of FLY

In [1]:
from fly import *
import os
import pickle
import math
import time

# SBPN Setting (S-box and Rotation off-set)

In [2]:


FLY_Sbox = ( 0x00, 0x9B, 0xC2, 0x15, 0x5D, 0x84, 0x4C, 0xD1, 0x67, 0x38, 0xEF, 0xB0, 0x7E, 0x2B, 0xF6, 0xA3 ,
             0xB9, 0xAA, 0x36, 0x78, 0x2F, 0x6E, 0xE3, 0xF7, 0x12, 0x5C, 0x9A, 0xD4, 0x89, 0xCD, 0x01, 0x45 ,
             0x2C, 0x63, 0x44, 0xDE, 0x02, 0x96, 0x39, 0x70, 0xBA, 0xE4, 0x18, 0x57, 0xA1, 0xF5, 0x8B, 0xCE ,
             0x51, 0x87, 0xED, 0xFF, 0xB5, 0xA8, 0xCA, 0x1B, 0xDF, 0x90, 0x6C, 0x32, 0x46, 0x03, 0x7D, 0x29 ,
             0xD5, 0xF2, 0x20, 0x5B, 0xCC, 0x31, 0x04, 0xBD, 0xA6, 0x41, 0x8E, 0x79, 0xEA, 0x9F, 0x68, 0x1C ,
             0x48, 0xE6, 0x69, 0x8A, 0x13, 0x77, 0x9E, 0xAF, 0xF3, 0x05, 0xCB, 0x2D, 0xB4, 0xD0, 0x37, 0x52 ,
             0xC4, 0x3E, 0x93, 0xAC, 0x40, 0xE9, 0x22, 0x56, 0x7B, 0x8D, 0xF1, 0x06, 0x17, 0x62, 0xBF, 0xDA ,
             0x1D, 0x7F, 0x07, 0xB1, 0xDB, 0xFA, 0x65, 0x88, 0x2E, 0xC9, 0xA5, 0x43, 0x58, 0x3C, 0xE0, 0x94 ,
             0x76, 0x21, 0xAB, 0xFD, 0x6A, 0x3F, 0xB7, 0xE2, 0xDD, 0x4F, 0x53, 0x8C, 0xC0, 0x19, 0x95, 0x08 ,
             0x83, 0xC5, 0x4E, 0x09, 0x14, 0x50, 0xD8, 0x9C, 0xF4, 0xEE, 0x27, 0x61, 0x3B, 0x7A, 0xA2, 0xB6 ,
             0xFE, 0xA9, 0x81, 0xC6, 0xE8, 0xBC, 0x1F, 0x5A, 0x35, 0x72, 0x99, 0x0A, 0xD3, 0x47, 0x24, 0x6D ,
             0x0B, 0x4D, 0x75, 0x23, 0x97, 0xD2, 0x60, 0x34, 0xC8, 0x16, 0xA0, 0xBB, 0xFC, 0xE1, 0x5E, 0x8F ,
             0xE7, 0x98, 0x1A, 0x64, 0xAE, 0x4B, 0x71, 0x85, 0x0C, 0xB3, 0x3D, 0xCF, 0x55, 0x28, 0xD9, 0xF0 ,
             0xB2, 0xDC, 0x5F, 0x30, 0xF9, 0x0D, 0x26, 0xC3, 0x91, 0xA7, 0x74, 0x1E, 0x82, 0x66, 0x4A, 0xEB ,
             0x6F, 0x10, 0xB8, 0xD7, 0x86, 0x73, 0xFB, 0x0E, 0x59, 0x2A, 0x42, 0xE5, 0x9D, 0xA4, 0x33, 0xC7 ,
             0x3A, 0x54, 0xEC, 0x92, 0xC1, 0x25, 0xAD, 0x49, 0x80, 0x6B, 0xD6, 0xF8, 0x0F, 0xBE, 0x7C, 0x11 )
fly_offset      = (7, 6, 5, 4, 3, 2, 1, 0)
NUM_SBOX_IN_A_STATE = 8
ALGNAME = "FLY"

K = SRBPN_cipher(FLY_Sbox, fly_offset, NUM_SBOX_IN_A_STATE, ALGNAME)

libastbb_srbn.Cal_Key_Recovery_Comp_11R.restype = c_double
libastbb_srbn.Cal_Key_Recovery_Comp_12R.restype = c_double


# Enumerate all input-output differential pairs of FLY cipher trails with probability ≥ 2^-63.
- Due to the file I/O structure, it runs slower on Windows and much faster on Linux.

In [3]:
fly_tar_rounds = 8

PROB_DICT = [0]
for r_idx in range(1, fly_tar_rounds+1):
    PROB_DICT = K.get_BDP(r_idx, PROB_DICT)
print(PROB_DICT)
PROB_DICT[-1] = -63.0
start_time = time.time()
K.get_all_BDT(fly_tar_rounds, PROB_DICT)
end_time = time.time()
elapsed_time = end_time - start_time

print(f"elapsed_time : {elapsed_time:.6f}sec")

[0.0, -4.0, -8.0, -16.0, -26.0, -34.0, -42.0, -50.0, -58.0]
elapsed_time : 714.849648sec


# Computation of key-recovery complexity using the distinguisher proposed in “Generating Graphs Packed with Paths” (Hall-Andersen & Vejre, FSE 2019), within our framework.


In [4]:
"""
Running the program provided in the paper “Generating Graphs Packed with Paths” yields 20 differential trails.  
Since the output changes each time the program runs, it can be inferred that there are more trails available internally,  
but only 20 are printed.  
Even if different trails are output, the key-recovery complexity remains the same,  
as they share the same activity pattern.
The following are the 20 differential trails obtained.
"""

differentials =[(0xd800000000000000,0x0044000000440000),
(0x0000e80000000000,0x0000004400000044),
(0x00000000008c0000,0x0000440000004400),
(0x000000000000008e,0x4400000044000000),
(0x00000000000000e8,0x4400000044000000),
(0x00e8000000000000,0x0000440000004400),
(0x00d8000000000000,0x0000440000004400),
(0x0000000000008d00,0x0000004400000044),
(0x00008d0000000000,0x0000004400000044),
(0x8f00000000000000,0x0044000000440000),
(0x0000000000008f00,0x0000004400000044),
(0x0000000000d80000,0x0000440000004400),
(0x0000008c00000000,0x4400000044000000),
(0x00000000000000f8,0x4400000044000000),
(0x00000000e8000000,0x0044000000440000),
(0x00c8000000000000,0x0000440000004400),
(0x000000000000008d,0x4400000044000000),
(0x00000000008d0000,0x0000440000004400),
(0x00000000000000c8,0x4400000044000000),
(0xe800000000000000,0x0044000000440000),]

# The results of the paper “Generating Graphs Packed with Paths” can also be verified using our tool as follows.
- GAP determines how many trails to cluster based on a given bound.
- For example, if GAP = 5 and the current differential bound is 2^-p,
- all individual trails with probability ≥ 2^(-p-5) will be clustered.

In [5]:
GAP = 5
fly_tar_rounds = 8

PROB_DICT = [0]
for r_idx in range(1, fly_tar_rounds+1):
    PROB_DICT = K.get_BDP(r_idx, PROB_DICT)
differentials_repre = []
for diff in differentials:
    x = (endian_change(diff[0]), list_to_int(K.INVPERM(int_to_list(endian_change(diff[1])))))
    differentials_repre.append(x)
    K.get_DT_IO(GAP, fly_tar_rounds, PROB_DICT, int_to_list(x[0]), int_to_list(x[1]))

In [6]:
import re
import math

def sum_round_probabilities(file_path):
    total_probability = 0.0
    probability_pattern = re.compile(r"\[ 8\] Round Best : 2\^\{(-?\d+\.\d+)}")

    with open(file_path, 'r') as file:
        for line in file:
            match = probability_pattern.search(line)
            if match:

                exponent = float(match.group(1))
                total_probability += 2 ** exponent

    return total_probability

for diff in differentials_repre:
    result=0
    xx = f"{format(diff[0],'016X')}"
    yy = f"{format(diff[1],'016X')}"
    file_path = f"FLY_DT_IO_R8_[{xx}-{yy}][58.0-63.0].txt"  # GAP = 5, diff_bound = -58
    result = sum_round_probabilities(file_path)
    print(K.cal_key_recov_comp_11R(diff[0], diff[1], -55.771, 1, 6.46))
    print(f"Total Probability: {math.log2(result)}")

98.79758420388566
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.714597781137755
90.87203482290003
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.714597781137755
98.79758420388566
Total Probability: -55.714597781137755
98.79758420388566
Total Probability: -55.714597781137755
98.79758420388566
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.77118130950412
106.7619701274411
Total Probability: -55.77118130950412
106.7619701274411
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.77118130950412
90.87203482290003
Total Probability: -55.77118130950412
106.7619701274411
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.714597781137755
90.87203482290003
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.77118130950412
98.79758420388566
Total Probability: -55.77

In [7]:
#(0x00000000000000c8, 0x4400000044000000)  
#This differential serves as the reference for identifying better distinguishers for our key-recovery attack.


In [8]:
NUM_OF_RIGHT_PAIR = 1 
c_e = 6.46 # The number of S-box evaluations per encryption

thres_comp = K.cal_key_recov_comp_11R(0x00000000000000c8, 0x4400000044000000, -55.771, 1, 6.46)
FLY_time_comp_dict = dict()
filename = f"{ALGNAME}_All_BDTs_R{fly_tar_rounds}.txt"

start_time = time.time()

print(f"start sorting {filename}...")   
with open(filename, 'r') as f:
    while True:

        line1 = f.readline()
        line2 = f.readline()
        


        if not line1 or not line2:
            break 

        line1 = line1.strip()
        line2 = line2.strip()

        if '|' not in line2:
            continue  

        try:
            hex_str, prob_str = map(str.strip, line2.split('|'))
            diff_in = int(line1, 16)
            diff_ou = int(hex_str, 16)
            proba   = float(prob_str)
            comp = K.cal_key_recov_comp_11R(diff_in, diff_ou, proba, NUM_OF_RIGHT_PAIR, c_e)
            if (comp <= thres_comp):
                FLY_time_comp_dict[(diff_in, diff_ou, proba)] = comp
        except ValueError:
            continue 

sorted_dict = sorted(FLY_time_comp_dict.items(), key = lambda item: item[1])
for idx in range(10):
    print(sorted_dict[idx])
sorted_FLY_time_comp_dict = dict(sorted_dict)
with open("sorted_comp_dict.pickle", "wb") as f:
    pickle.dump(sorted_FLY_time_comp_dict, f)

end_time = time.time()
elapsed_time = end_time - start_time

print(f"elapsed_time : {elapsed_time:.6f}sec")


start sorting FLY_All_BDTs_R8.txt...
((2594073385365405696, 1224979098929987584, -58.0), 80.93258752505177)
((4755801206503243776, 1224979098929987584, -58.0), 80.93258752505177)
((2594073385365405696, 4683743612800860160, -58.0), 81.9226287107821)
((2594073385365405696, 1441151881849077760, -58.0), 81.9226287107821)
((4755801206503243776, 4683743612800860160, -58.0), 81.9226287107821)
((4755801206503243776, 1441151881849077760, -58.0), 81.9226287107821)
((2594073385365405696, 18691697676544, -60.0), 82.93258752505177)
((4755801206503243776, 18691697676544, -60.0), 82.93258752505177)
((2594073385365405696, 71468255810560, -60.0), 83.9226287107821)
((2594073385365405696, 21990232572160, -60.0), 83.9226287107821)
elapsed_time : 31.762121sec


# (2594073385365405696, 1224979098929987584) = (0x2400000000000000, 0x1100000011000000)

# Perform clustering on the identified key-recovery-attack friendly distinguishers.


In [9]:
opt_dist = (0x2400000000000000, 0x1100000011000000)
NUM_OF_RIGHT_PAIR = 4 # To ensure a success probability above 90%, we use 4 right pairs.
c_e1 = 6.46 # 11R FLY
c_e2 = 6.58 # 12R FLY
GAP = 17
fly_tar_rounds = 8

PROB_DICT = [0]
for r_idx in range(1, fly_tar_rounds+1):
    PROB_DICT = K.get_BDP(r_idx, PROB_DICT)
#K.get_DT_IO(GAP, fly_tar_rounds, PROB_DICT, int_to_list(0x2400000000000000), int_to_list(0x1100000011000000))

result=0
xx = f"{format(opt_dist[0],'016X')}"
yy = f"{format(opt_dist[1],'016X')}"
file_path = f"FLY_DT_IO_R8_[{xx}-{yy}][58.0-75.0].txt"  # GAP = 17, diff_bound = -58
result = sum_round_probabilities(file_path)
print(f"Total Probability: {math.log2(result)}")
print(K.cal_key_recov_comp_11R(opt_dist[0], opt_dist[1], -56.77, NUM_OF_RIGHT_PAIR, c_e1))
print(K.cal_key_recov_comp_12R(opt_dist[0], opt_dist[1], -56.77, NUM_OF_RIGHT_PAIR, c_e2))



Total Probability: -56.77548834662345
81.70258752505177
118.92928411166672


# The LaTeX table reveals the pattern of the distinguisher we used.

In [10]:
K = Latex_tab(FLY_Sbox, fly_offset, NUM_SBOX_IN_A_STATE, ALGNAME)

In [11]:
from fly_latex_tabgen import latex_gen_inter_value

In [12]:
opt_dist = (0x2400000000000000, 0x1100000011000000)

In [13]:
_1_mask = list_to_int(K.INVPERM(int_to_list(opt_dist[0])))
_0_mask = 0xffffffffffffffff ^ _1_mask
_d_mask = 0xffffffffffffffff ^ _1_mask ^ _0_mask

_1_mask = int_to_list(_1_mask)
_0_mask = int_to_list(_0_mask)
_d_mask = int_to_list(_d_mask)

_1_mask_IN = opt_dist[0]
_0_mask_IN = 0xffffffffffffffff ^ _1_mask_IN
_d_mask_IN = 0xffffffffffffffff ^ _1_mask_IN ^ _0_mask_IN

_1_mask_IN = int_to_list(_1_mask_IN)
_0_mask_IN = int_to_list(_0_mask_IN)
_d_mask_IN = int_to_list(_d_mask_IN)


_1_mask_OU = opt_dist[1]
_0_mask_OU = 0xffffffffffffffff ^ list_to_int(K.PERM(int_to_list(opt_dist[1])))
_d_mask_OU = 0xffffffffffffffff ^ _1_mask_OU ^ _0_mask_OU

_1_mask_OU = int_to_list(_1_mask_OU)
_0_mask_OU = int_to_list(_0_mask_OU)
_d_mask_OU = int_to_list(_d_mask_OU)

In [14]:
pre_str = "%"*30 + "FLY 11-Round" + "%"*30
print(pre_str)

lines =[K.Cal_INV_1r_BP(int_to_list(opt_dist[0])),K.bitmask_to_string(_0_mask, _1_mask, _d_mask),K.bitmask_to_string(_0_mask_IN, _1_mask_IN, _d_mask_IN),K.bitmask_to_string(_0_mask_IN, _1_mask_IN, _d_mask_IN), 
        K.bitmask_to_string(_0_mask_OU, _1_mask_OU, _d_mask_OU), K.Cal_1r_BP(int_to_list(opt_dist[1])), K.Cal_BP_1r_BP(int_to_list(opt_dist[1])), K.Cal_BP_1r_BP(int_to_list(opt_dist[1])), K.Cal_2r_BP(int_to_list(opt_dist[1])), K.Cal_BP_2r_BP(int_to_list(opt_dist[1])), K.Cal_BP_2r_BP(int_to_list(opt_dist[1]))]

latex_gen_inter_value(lines)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%FLY 11-Round%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{table}[h]
\centering
\scriptsize
\caption{Difference Values of $\Delta K^i$, $\Delta S^i$, and $\Delta R^i$ used in the key-recovery attack on 11-Round \textsf{FLY}.}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|}
\hline
 & 0-th col. & 1-st col. & 2-nd col. & 3-rd col. & 4-th col. & 5-th col. & 6-th col. & 7-th col. \\
\hline
$\Delta K^0$ & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} \\
$\Delta S^1$ & \texttt{00000000} & \texttt{00000000} & \texttt{00100000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000100} & \texttt{00000000} & \texttt{00000000} \\
$\Delta R^1$ & \texttt{00100100} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^1$ & \texttt{00100100} & \texttt{00000000} & \texttt{00000000} 

In [15]:
latex_code = r"""
\documentclass{article}
\usepackage[a4paper, landscape, margin=1in]{geometry}
\usepackage{array}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{booktabs}

\begin{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%FLY 11-Round%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{table}[h]
\centering
\scriptsize
\caption{Difference Values of $\Delta K^i$, $\Delta S^i$, and $\Delta R^i$ used in the key-recovery attack on 11-Round \textsf{FLY}.}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|}
\hline
 & 0-th col. & 1-st col. & 2-nd col. & 3-rd col. & 4-th col. & 5-th col. & 6-th col. & 7-th col. \\
\hline
$\Delta K^0$ & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} \\
$\Delta S^1$ & \texttt{00000000} & \texttt{00000000} & \texttt{00100000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000100} & \texttt{00000000} & \texttt{00000000} \\
$\Delta R^1$ & \texttt{00100100} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^1$ & \texttt{00100100} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^9$ & \texttt{00010001} & \texttt{000*000*} & \texttt{00000000} & \texttt{00000000} & \texttt{00010001} & \texttt{000*000*} & \texttt{00000000} & \texttt{00000000} \\
$\Delta S^{11}$ & \texttt{00000000} & \texttt{*01**01*} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{*01**01*} & \texttt{00000000} & \texttt{00000000} \\
$\Delta R^{11}$ & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} \\
$\Delta K^{11}$ & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} \\
$\Delta S^{12}$ & \texttt{00000000} & \texttt{********} & \texttt{********} & \texttt{********} & \texttt{00000000} & \texttt{********} & \texttt{********} & \texttt{********} \\
$\Delta R^{12}$ & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} \\
$\Delta K^{12}$ & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} \\
\hline
\end{tabular}
\label{tab:difftrail64}
\end{table}

\end{document}
"""
with open("fly_table_11r.tex", "w") as f:
    f.write(latex_code)

In [16]:
import subprocess

result = subprocess.run(["pdflatex", "fly_table_11r.tex"], capture_output=True, text=True)


if result.returncode == 0:
    print("PDF successfully generated.")
else:
    print("Error during PDF generation:")
    print(result.stderr)


PDF successfully generated.


In [17]:
from IPython.display import IFrame

IFrame("fly_table_11r.pdf", width=800, height=600)


In [18]:

pre_str = "%"*30 + "FLY 12-Round" + "%"*30
print(pre_str)
lines =[K.Cal_INV_2r_BP(int_to_list(opt_dist[0])), K.Cal_INV_BP_1r_BP(int_to_list(opt_dist[0])), K.Cal_INV_1r_BP(int_to_list(opt_dist[0])), K.Cal_INV_1r_BP(int_to_list(opt_dist[0])),K.bitmask_to_string(_0_mask, _1_mask, _d_mask),K.bitmask_to_string(_0_mask_IN, _1_mask_IN, _d_mask_IN),K.bitmask_to_string(_0_mask_IN, _1_mask_IN, _d_mask_IN), 
        K.bitmask_to_string(_0_mask_OU, _1_mask_OU, _d_mask_OU), K.Cal_1r_BP(int_to_list(opt_dist[1])), K.Cal_BP_1r_BP(int_to_list(opt_dist[1])), K.Cal_BP_1r_BP(int_to_list(opt_dist[1])), K.Cal_2r_BP(int_to_list(opt_dist[1])), K.Cal_BP_2r_BP(int_to_list(opt_dist[1])), K.Cal_BP_2r_BP(int_to_list(opt_dist[1]))]

latex_gen_inter_value(lines)



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%FLY 12-Round%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{table}[h]
\centering
\scriptsize
\caption{Difference Values of $\Delta K^i$, $\Delta S^i$, and $\Delta R^i$ used in the key-recovery attack on 12-Round \textsf{FLY}.}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|}
\hline
 & 0-th col. & 1-st col. & 2-nd col. & 3-rd col. & 4-th col. & 5-th col. & 6-th col. & 7-th col. \\
\hline
$\Delta K^0$ & \texttt{00010010} & \texttt{00001001} & \texttt{10000100} & \texttt{01000010} & \texttt{00100001} & \texttt{10010000} & \texttt{01001000} & \texttt{00100100} \\
$\Delta S^1$ & \texttt{00010010} & \texttt{00001001} & \texttt{10000100} & \texttt{01000010} & \texttt{00100001} & \texttt{10010000} & \texttt{01001000} & \texttt{00100100} \\
$\Delta R^1$ & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^1$ & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} 

In [19]:
latex_code = r"""
\documentclass{article}
\usepackage[a4paper, landscape, margin=1in]{geometry}
\usepackage{array}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{booktabs}

\begin{document}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%FLY 12-Round%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{table}[h]
\centering
\scriptsize
\caption{Difference Values of $\Delta K^i$, $\Delta S^i$, and $\Delta R^i$ used in the key-recovery attack on 12-Round \textsf{FLY}.}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|}
\hline
 & 0-th col. & 1-st col. & 2-nd col. & 3-rd col. & 4-th col. & 5-th col. & 6-th col. & 7-th col. \\
\hline
$\Delta K^0$ & \texttt{00010010} & \texttt{00001001} & \texttt{10000100} & \texttt{01000010} & \texttt{00100001} & \texttt{10010000} & \texttt{01001000} & \texttt{00100100} \\
$\Delta S^1$ & \texttt{00010010} & \texttt{00001001} & \texttt{10000100} & \texttt{01000010} & \texttt{00100001} & \texttt{10010000} & \texttt{01001000} & \texttt{00100100} \\
$\Delta R^1$ & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^1$ & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} & \texttt{11111111} & \texttt{00000000} & \texttt{00000000} \\
$\Delta S^2$ & \texttt{00000000} & \texttt{00000000} & \texttt{00100000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000100} & \texttt{00000000} & \texttt{00000000} \\
$\Delta R^2$ & \texttt{00100100} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^2$ & \texttt{00100100} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} \\
$\Delta K^{10}$ & \texttt{00010001} & \texttt{000*000*} & \texttt{00000000} & \texttt{00000000} & \texttt{00010001} & \texttt{000*000*} & \texttt{00000000} & \texttt{00000000} \\
$\Delta S^{11}$ & \texttt{00000000} & \texttt{*01**01*} & \texttt{00000000} & \texttt{00000000} & \texttt{00000000} & \texttt{*01**01*} & \texttt{00000000} & \texttt{00000000} \\
$\Delta R^{11}$ & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} \\
$\Delta K^{11}$ & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} & \texttt{00000000} & \texttt{*000*000} & \texttt{000*000*} & \texttt{00100010} \\
$\Delta S^{12}$ & \texttt{00000000} & \texttt{********} & \texttt{********} & \texttt{********} & \texttt{00000000} & \texttt{********} & \texttt{********} & \texttt{********} \\
$\Delta R^{12}$ & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} \\
$\Delta K^{12}$ & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} & \texttt{0***0***} & \texttt{***0***0} & \texttt{**0***0*} & \texttt{*0***0**} \\
\hline
\end{tabular}
\label{tab:difftrail64}
\end{table}

\end{document}
"""
with open("fly_table_12r.tex", "w") as f:
    f.write(latex_code)

In [20]:
import subprocess

result = subprocess.run(["pdflatex", "fly_table_12r.tex"], capture_output=True, text=True)


if result.returncode == 0:
    print("PDF successfully generated.")
else:
    print("Error during PDF generation:")
    print(result.stderr)


PDF successfully generated.


In [21]:
from IPython.display import IFrame

IFrame("fly_table_12r.pdf", width=800, height=600)


# The tables and computational complexities presented in our paper can be verified both through manual calculations and by comparing them with the outputs generated by our program.

In [22]:
import math
def summation(L, i):
    sum = 0
    for j in range(i+1):
        sum = sum + L[j]
    return sum
def cal_inner_kr_comp(f,g):
    diff_list = [f[i] - g[i] for i in range(len(f))]
    sum = 0
    for i in range(len(f)):
        sum = sum + 2**summation(diff_list,i)
        #print(sum)
        
    return sum

In [23]:
tab_g_fly_11r = [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 6, 6]
tab_f_fly_11r = [8, 8, 8, 8, 6, 6, 6, 6, 0, 0, 4, 4]

In [24]:
print(math.log2(4*2**(56.77+1+16+48-64)*(1 + 2**(-6.46)*cal_inner_kr_comp(tab_g_fly_11r,tab_f_fly_11r)) + 4*2**(56.77+1) + 2**(128-58)))

81.70302067766097


In [25]:
tab_g_fly_12r = [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 0, 0, 0, 0, 0, 0, 0, 0, 6, 6]
tab_f_fly_12r = [6, 6, 6, 6, 6, 6, 6, 6, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 4, 4]

In [26]:
print(math.log2(4*2**(56.77+1+64+48-64)*(1 + 2**(-6.58)*cal_inner_kr_comp(tab_g_fly_12r,tab_f_fly_12r)) + 2**(61.89) + 2**(128-25)))

118.92930715237281


# The key-recovery complexity and the infeasibility of the 12-round key-recovery attack, as presented in the paper Generating Graphs Packed with Paths, are as follows:

In [28]:
tab_g_fly_11r_ref = [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 5, 5, ]
tab_f_fly_11r_ref = [8, 8, 8, 6, 6, 6, 6, 6, 6, 6, 6, 8, 8]

In [29]:
print(math.log2(4*2**(55.77+1+24+64-64)*(1 + 2**(-6.46)*cal_inner_kr_comp(tab_g_fly_11r_ref,tab_f_fly_11r_ref)) + 2**(59.77) + 2**(128-58)))

92.87103501112121


In [27]:
print(K.cal_key_recov_comp_11R(0x00000000000000c8, 0x4400000044000000, -55.77, 4, 6.46))

92.87103482290004


In [30]:
print(K.cal_key_recov_comp_12R(0x00000000000000c8, 0x4400000044000000, -55.77, 4, 6.58))

142.24614503260992
